examples/custom-congestion-controller/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+    /// For this example, we track the size of the congestion window in bytes and
+    /// the number of bytes in flight.
+    #[derive(Debug, Clone)]
+    pub struct MyCongestionController {
+        congestion_window: u32,
+        bytes_in_flight: u32,
+    }
+
+    /// The following is a simple implementation of the `CongestionController` trait
+    /// that increases the congestion window by the number of bytes acknowledged and
+    /// decreases the congestion window by half when packets are lost.
+    #[allow(unused)]
+    impl CongestionController for MyCongestionController {
+        type PacketInfo = ();
+
+        fn congestion_window(&self) -> u32 {
+            self.congestion_window
+        }
+
+        fn bytes_in_flight(&self) -> u32 {
+            self.bytes_in_flight
+        }
+
+        fn is_congestion_limited(&self) -> bool {
+            self.congestion_window < self.bytes_in_flight
+        }
+
+        fn requires_fast_retransmission(&self) -> bool {
+            false
+        }
+
+        fn on_packet_sent<Pub: Publisher>(
+            &mut self,
+            time_sent: Timestamp,
+            sent_bytes: usize,
+            app_limited: Option<bool>,
+            rtt_estimator: &RttEstimator,
+            publisher: &mut Pub,
+        ) -> Self::PacketInfo {
+            self.bytes_in_flight += sent_bytes as u32;
+        }
+
+        fn on_rtt_update<Pub: Publisher>(
+            &mut self,
+            time_sent: Timestamp,
+            now: Timestamp,
+            rtt_estimator: &RttEstimator,
+            publisher: &mut Pub,
+        ) {
+            // no op
+        }
+
+        fn on_ack<Pub: Publisher>(
+            &mut self,
+            newest_acked_time_sent: Timestamp,
+            bytes_acknowledged: usize,
+            newest_acked_packet_info: Self::PacketInfo,
+            rtt_estimator: &RttEstimator,
+            random_generator: &mut dyn RandomGenerator,
+            ack_receive_time: Timestamp,
+            publisher: &mut Pub,
+        ) {
+            self.bytes_in_flight -= bytes_acknowledged as u32;
+            self.congestion_window += bytes_acknowledged as u32;
+        }
+
+        fn on_packet_lost<Pub: Publisher>(
+            &mut self,
+            lost_bytes: u32,
+            packet_info: Self::PacketInfo,
+            persistent_congestion: bool,
+            new_loss_burst: bool,
+            random_generator: &mut dyn RandomGenerator,
+            timestamp: Timestamp,
+            publisher: &mut Pub,
+        ) {
+            self.bytes_in_flight -= lost_bytes;
+            self.congestion_window = (self.congestion_window as f32 * 0.5) as u32;
+        }
+
+        fn on_explicit_congestion<Pub: Publisher>(
+            &mut self,
+            ce_count: u64,
+            event_time: Timestamp,
+            publisher: &mut Pub,
+        ) {
+            self.congestion_window = (self.congestion_window as f32 * 0.5) as u32;
```

Comment:
I know it's an example but this essentially means if you get any loss ever you'll never recover the cwnd. Not sure if there's a more concise impl that gives an idea of the appropriate reactions?

Reply:
Yeah, the 0.5 was from New Reno, though this would be much more severe than New Reno as a burst of lost packets would continue reducing the cwnd each time, when the real New Reno would enter "recovery" and only reduce the cwnd once until it exits recovery. I think that is a little more detail than I want to put in this example, but I can improve the comments a bit to emphasize that an actual congestion controller would be more complex